controllers/snapshot/snapshot_adapter.go

```diff
@@ -188,44 +193,92 @@ TestScenarioLoop:
+					err = a.labelBindingWithScenario(binding, integrationTestScenario.Name)
+					if err != nil {
+						a.logger.Error(err, "Failed to update snapshotEnvironmentbinding label",
+							"binding.Name", binding.Name,
+							gitops.SnapshotTestScenarioLabel, integrationTestScenario.Name)
+						return reconciler.RequeueWithError(err)
+					}
+					a.logger.LogAuditEvent("Updated snapshotEnvironmentbinding label", binding, h.LogActionUpdate,
+						gitops.SnapshotTestScenarioLabel, integrationTestScenario.Name)
+				} else {
+					a.logger.Info("SnapshotEnvironmentBinding already contains label",
+						"binding.Name", binding.Name,
+						"integrationScenario.Name", integrationTestScenario.Name)
+				}
 				continue TestScenarioLoop
 			}
 		}
 
 		//get the existing environment according to environment name from integrationTestScenario
 		existingEnv, err := a.getEnvironmentFromIntegrationTestScenario(&integrationTestScenario)
 		if err != nil {
-			a.logger.Error(err, "Failed to find the env defined in integrationTestScenario.",
+			a.logger.Error(err, "Failed to find the env defined in integrationTestScenario",
 				"integrationTestScenario.Namespace", integrationTestScenario.Namespace,
 				"integrationTestScenario.Name", integrationTestScenario.Name)
-			return reconciler.RequeueOnErrorOrStop(err)
+			return reconciler.RequeueWithError(err)
 		}
 
 		//create an ephemeral copy env of existing environment
 		copyEnv, err := a.createCopyOfExistingEnvironment(existingEnv, a.snapshot.Namespace, &integrationTestScenario, a.snapshot, a.application)
 
 		if err != nil {
-			a.logger.Error(err, "Copying of environment failed.")
-			return reconciler.RequeueOnErrorOrStop(err)
+			a.logger.Error(err, "Copying of environment failed")
+			return reconciler.RequeueWithError(err)
 		}
 		a.logger.LogAuditEvent("An ephemeral Environment is created for integrationTestScenario",
 			copyEnv, h.LogActionAdd,
 			"integrationTestScenario.Name", integrationTestScenario.Name)
 
-		components, err := a.loader.GetAllApplicationComponents(a.client, a.context, a.application)
-		if err != nil {
-			return reconciler.RequeueWithError(err)
-		}
-
 		binding, err := a.createSnapshotEnvironmentBindingForSnapshot(a.application, copyEnv, a.snapshot, components)
 		if err != nil {
-			a.logger.Error(err, "Failed to create snapshotEnvironmentbinding for snapshot.",
-				"binding", binding,
+			a.logger.Error(err, "Failed to create snapshotEnvironmentbinding for snapshot",
+				"snapshot", a.snapshot.Name,
+				"environment.Name", copyEnv.Name,
 				"snapshot.Spec.Components", a.snapshot.Spec.Components)
+			return reconciler.RequeueWithError(err)
 		}
 		a.logger.LogAuditEvent("A snapshotEnvironmentbinding is created", binding, h.LogActionAdd,
 			"environment.Name", copyEnv.Name,
 			"integrationTestScenario.Name", integrationTestScenario.Name)
+
+		err = a.labelBindingWithScenario(binding, integrationTestScenario.Name)
```

Comment:
The code as it is now does work, and the possible errors are handled well.

I'm thinking this might be simplified further to just add the label as part of creating the SnapshotEnvironmentBinding. It would require modifying the `createSnapshotEnvironmentBindingForSnapshot` function to take the optional label name and value (or something similar to that) and add it before creating the resource on the cluster.
@hongweiliu17 wdyt about it?  I'm fine with keeping it this way too if you think it's not worth the hassle to implement it that way.

Reply:
Thank you. I agree. I thought adding it to `createSnapshotEnvironmentBindingForSnapshot` is simpler, but  I found `createSnapshotEnvironmentBindingForSnapshot` is called [here](https://github.com/redhat-appstudio/integration-service/blob/main/controllers/snapshot/snapshot_adapter.go#L348) and it doesn't need this label. You are right, we may add optional label name and value params.

Let me update it.

Reply:
Updated. Thanks.

Reply:
Looks good!